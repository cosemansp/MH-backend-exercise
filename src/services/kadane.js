/* eslint-disable prefer-destructuring */
export function kadanes(array) {
  const n = array.length;
  const dp = [];

  // base condition
  dp[0] = array[0];
  let answer = 0;
  for (let i = 1; i < n; i += 1) {
    dp[i] = Math.max(dp[i - 1], 0) + array[i];
    answer = Math.max(answer, dp[i]);
  }
  return answer;
}

export function sumArray(a1, a2) {
  return a1.map((num, idx) => num + a2[idx]);
}

// see also https://blog.abelotech.com/posts/array-conversion-2-dimensional-javascript/
export function splitArray(array, howMany) {
  const result = [];
  const input = [...array];
  while (input[0] !== undefined) {
    const part = input.splice(0, howMany);
    result.push(part);
  }
  return result;
}

export function kadanes2d(array) {
  let top = 0;
  let bottom = 0;
  let cache = [...array[top]];
  let maxSum = Math.max(0, kadanes(cache));

  while (bottom < array.length - 1 && top < array.length - 1) {
    bottom += 1;
    cache = sumArray(cache, array[bottom]);
    maxSum = Math.max(maxSum, kadanes(cache));
    if (bottom === array.length - 1) {
      top += 1;
      bottom = top;
      cache = [...array[top]];
    }
  }

  return maxSum;
}
