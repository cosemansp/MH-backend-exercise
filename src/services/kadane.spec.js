import { kadanes, kadanes2d, splitArray, sumArray } from './kadane';

describe('Kadane', () => {
  test('1d array', () => {
    expect(kadanes([2, 0, 2, -3])).toBe(4);
    expect(kadanes([5, 1, -2, -3])).toBe(6);
    expect(kadanes([0, 1, -1, -3])).toBe(1);
    expect(kadanes([1, 14, -2, 4])).toBe(17);
    expect(kadanes([1, 6, -2, 3])).toBe(8);
  });

  test('2d array - [2, 0], [0, 1], [0, -2]', () => {
    expect(kadanes2d([[2, 0], [0, 1], [0, -2]])).toBe(3);
  });

  test('2d array - [2, 0, -3], [0, 1, 2], [0, 0, 4]', () => {
    expect(kadanes2d([[2, 0, -3], [0, 1, 2], [0, 0, 4]])).toBe(7);
  });

  test('split array', () => {
    expect(splitArray([2, 0, 2, -3], 2)).toEqual([[2, 0], [2, -3]]);
    expect(splitArray([0, 0, 2, -3], 2)).toEqual([[0, 0], [2, -3]]);
    expect(splitArray([2, 0, 2, -3, 5, 1], 3)).toEqual([[2, 0, 2], [-3, 5, 1]]);
    expect(splitArray([0, -2, -7, 0, 9, 2, -6, 2, -4, 1, -4, 1, -1, 8, 0, -2], 4)).toMatchSnapshot();
  });

  test('sum array', () => {
    const result = sumArray([2, 0], [2, -3]);
    expect(result).toEqual([4, -3]);
  });

  test('combined', () => {
    const array = splitArray([0, -2, -7, 0, 9, 2, -6, 2, -4, 1, -4, 1, -1, 8, 0, -2], 4);
    expect(kadanes2d(array)).toBe(15);
  });
});
