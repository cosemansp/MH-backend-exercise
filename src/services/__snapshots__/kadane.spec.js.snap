// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kadane split array 1`] = `
Array [
  Array [
    0,
    -2,
    -7,
    0,
  ],
  Array [
    9,
    2,
    -6,
    2,
  ],
  Array [
    -4,
    1,
    -4,
    1,
  ],
  Array [
    -1,
    8,
    0,
    -2,
  ],
]
`;
